### PR TITLE
[JIT] SubgraphRewriter: Rename pattern fusion to subgraph rewrite.

### PR DESCRIPTION
**Pull request by ZolotukhinM**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #20084 [JIT] SubgraphRewriter: Add a support for arbitrary replacement graphs in subgraph rewriter.
* #20083 [JIT] SubgraphRewriter: Expose runOnGraph and use it in tests.
* **#20082 [JIT] SubgraphRewriter: Rename pattern fusion to subgraph rewrite.**

Differential Revision: [D15190193](https://our.internmc.facebook.com/intern/diff/D15190193)

### CODE REVIEW

**Comment by raghuramank100** [comment]
Feature request: 

We will run into cases, where we want to do more than a pattern replacement. Example: Consider replacing a  dequant->conv->quant with a qconv. The weight_scale, input_scale and output_scale are nodes in the graph, which we can access directly.

 We could (not at present) have a case where we will need to do a computation, requant_scale =round(output_scale/(weight_scale*input_scale)*2^k) and provide this as an input to a qconv(x,w,requant_scale) which would internally do: round(conv(xq,wq)*requant_scale>>k). How do we plan to handle cases like this?  Would we have a 'pack/init_conv' op that would do the necessary pre-processing so that the required values (like requant_scale) are available in IR?

**Review by raghuramank100 — APPROVED** [review]

File: torch/csrc/jit/passes/subgraph_rewrite.h [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+/** \brief Run pattern-based subgraph rewrites on all methods in the module.
+ *
+ * This pass will go through all methods in the module and try to replace all
+ * recognized patterns (see SubgraphRewriter::RegisterDefaultPatterns for the
+ * list of these patterns).
+ */
+TORCH_API std::shared_ptr<script::Module> PatternBasedRewrite(
+    std::shared_ptr<script::Module> module);
+
+/** \brief A class implementing API for pattern-based subgraph rewrites.
```
Comment (suo) [inline torch/csrc/jit/passes/subgraph_rewrite.h]: random style nit: I think `\brief` makes the comment less readable. For doxygen, you can configure so the first line in a javadoc-style comment is automatically taken as the brief description anyway.

So the comment could look like
```
/**
 * this is the brief description.
 * This is the rest
 */

**Comment by ZolotukhinM** [comment]
> We will run into cases, where we want to do more than a pattern replacement. Example: Consider replacing a dequant->conv->quant with a qconv. The weight_scale, input_scale and output_scale are nodes in the graph, which we can access directly.
> We could (not at present) have a case where we will need to do a computation, requant_scale =round(output_scale/(weight_scale*input_scale)*2^k) and provide this as an input to a qconv(x,w,requant_scale) which would internally do: round(conv(xq,wq)*requant_scale>>k). How do we plan to handle cases like this? Would we have a 'pack/init_conv' op that would do the necessary pre-processing so that the required values (like requant_scale) are available in IR?

There are several options: 1) We could represent this computation directly in IR, in that case everything would just work as-is. 2) This pass is not intended to replace the entire compiler. When there is something it could not reasonably do, we always could add another pass. 3) We could add a preparation pass that would precompute these values and insert the corresponding values into IR.

In the particular case you mentioned, I think we would do (3).

**Review by zdevito — APPROVED** [review]
Assuming this is just a rename, I didn't review the code otherwise.

**Pull request closed without merging**